### PR TITLE
Fix #3662: Complete lightweight checkpointing

### LINKED ISSUE #3479: Enable checkpointing once mechanism to resume exploration with checkpoints is implemented.
Checkpointing is disabled until the mechanism to resume explorations with checkpoints is implemented. 
Once resuming mechanism is complete checkpointing will have to be enabled by changing the value of the boolean shouldSavePartialProgress in the following classes:
1. TopicLessonsFragment.kt
2. RecentlyPlayedFragmentPresenter.kt
3. StoryChapterSummaryViewModel.kt

### LINKED ISSUE #3662: Stop marking lessons as started_not_completed once checkpointing is enabled.
Once checkpointing is enabled, remove the function that marks lessons as started_not_completed and the tests that test this behaviour.

### PR DESCRIPTION
**Pull request by MaskedCarrot**
## Explanation

Fixes #3662

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [ ] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### Final Result

https://user-images.githubusercontent.com/53645584/130246749-7c139a11-a806-4683-b3bc-f5717ade88aa.mp4

#### StateFragmentTest
Screenshot from 2021-08-21 04-19-07

#### ExplorationActivityTest
Screenshot from 2021-08-20 03-04-28

### CODE REVIEW

File: domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
Comment (BenHenning) [inline domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt]: Is this test/test name correct? It seems contradictory (& has a capitalization typo).

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Nothing is actually being changed here--how is the feature being enabled?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MaskedCarrot. We ought to get a review from Srushti that the feature as a whole is implemented correctly, but given the deadline I think it's reasonable to perform that check after GSoC ends.

Otherwise, had a couple of comments. PTAL.

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt
Comment (MaskedCarrot) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: checkpointing was enabled in PR (#3672), I have linked that PR with this issue.

File: domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt
Comment (MaskedCarrot) [reply domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt]: this test is testing the ongoing list when one chapter is marked as in progress saved and another is marked as in progress not saved. I have updated the test name to `fun testStoryList_markLessonInProgressSaved_anotherLessonInProgressNotSaved_ongoingListCorrect()` to clearly indicate this.

Comment (BenHenning) [reply domain/src/test/java/org/oppia/android/domain/topic/TopicListControllerTest.kt]: Ah, that's much clearer--thanks!

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Ah, we actually have a TODO check in place now that will reopen issues if they're closed with TODOs still checked into the codebase. I actually suggest that you remove the TODOs in the same PR that's resolving them since that's what's normally expected.

**Review by BenHenning — COMMENTED** [review]
Thanks @MaskedCarrot. I think this PR LGTM, but will wait to approve until it's based on develop.

Also, followed up to 1 open comment thread--PTAL.

**Review by rt4914 — APPROVED** [review]
LGTM for codeowner file.

File: app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt
Comment (BenHenning) [reply app/src/main/java/org/oppia/android/app/home/recentlyplayed/RecentlyPlayedFragmentPresenter.kt]: Looks resolved now.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @MaskedCarrot. LGTM. Please reassign once the blocking PR is merged & this is brought up-to-date with passing CI.

**Comment by BenHenning** [comment]
Though, @MaskedCarrot it seems there are failing tests. Please address either here or an earlier PR in the chain (depending on where they originated).

**Comment by MaskedCarrot** [comment]
@BenHenning  PTAL
The tests that retrieve checkpoints are failing on Bazel and passing on Gradle. 
This is happening because before retrieving the checkpoint, we match the saved checkpoint version with the current exploration, and since we load exploration from text proto on Bazel, the version doesn't match (because the value of "version"  in textproto of exploration is 0).

**Comment by BenHenning** [comment]
@MaskedCarrot then we may need to change the tests--these need to pass on both Bazel & Gradle. We can't be checking in code that breaks the CI environment.

**Review by anandwana001 — COMMENTED** [review]
Assign me back once test passes on CI.

**Comment by BenHenning** [comment]
@MaskedCarrot per my earlier comment, this isn't yet based on develop with passing CI. Please reassign at that point.

**Comment by anandwana001** [comment]
> Assign me back once test passes on CI.

Following on this, assign me back once this PR get develop branch as base branch.

**Review by anandwana001 — APPROVED** [review]
Code owner file LGTM

**Comment by vinitamurthi** [comment]
I think I was added by mistake, I dont codeown any of these files. Removing myself!

**Review by BenHenning — APPROVED** [review]
Thanks @MaskedCarrot. LGTM!

**Comment by BenHenning** [comment]
Given everyone's approved, CI is passing, and there are not outstanding conversation threads, merging this.

**Pull request merged**
